_posts/2015-07-23-postgres-review-operators.md

```diff
@@ -0,0 +1,292 @@
+ t
+(1 row)
+{% endhighlight %}
+
++ There is no operator if you need an **intersection**, but we can achieve the
+same result using:
+
+{% highlight sql %}
+SELECT array(
+  SELECT UNNEST(array[3,6,7,8])
+  INTERSECT
+  SELECT UNNEST(array[4,6,7,9])
+);
+
+ array
+-------
+ {6,7}
+(1 row)
+{% endhighlight %}
+
+We can use **UNION** instead of the **INTERSECT** to join elements from each
+Array without duplicates.
+
+### JSON
+
+There are three kinds of operators and each one works with arrays and
+key/values.
+
++ `->` Returns an array element by index `OR` a value by key as **JSON**.
+
+{% highlight sql %}
+SELECT '[{"name": "bernardo"},{"country": "brazil"}]'::json->0;
+
+       ?column?
+----------------------
+ {"name": "bernardo"}
+(1 row)
+
+SELECT pg_typeof(
+  '[{"name": "bernardo"},{"country": "brazil"}]'::json->0
+);
+
+ pg_typeof
+-----------
+ json
+(1 row)
+
+SELECT '{"name": "bernardo", "country": "brazil"}'::json->'name';
+
+  ?column?
+------------
+ "bernardo"
+(1 row)
+
+SELECT pg_typeof(
+  '{"name": "bernardo", "country": "brazil"}'::json->'name'
+);
+
+ pg_typeof
+-----------
+ json
+(1 row)
+{% endhighlight %}
+
+Since `->` returns a **JSON** we can chain this operator:
+
+{% highlight sql %}
+SELECT '{"country": {"name": "brazil"} }'::json->'country'->'name';
+
+?column?
+----------
+ "brazil"
+(1 row)
+{% endhighlight %}
+
++ `->>` Returns an array element by index `OR` a value by key as **TEXT**.
+
+{% highlight sql %}
+SELECT '[{"name": "bernardo"},{"country": "brazil"}]'::json->>0;
+
+       ?column?
+----------------------
+ {"name": "bernardo"}
+(1 row)
+
+SELECT pg_typeof(
+  '[{"name": "bernardo"},{"country": "brazil"}]'::json->>0
+);
+
+ pg_typeof
+-----------
+ text
+(1 row)
+
+SELECT '{"name": "bernardo", "country": "brazil"}'::json->>'name';
+
+  ?column?
+------------
+ bernardo
+(1 row)
+
+SELECT pg_typeof(
+  '{"name": "bernardo", "country": "brazil"}'::json->>'name'
+);
+
+ pg_typeof
+-----------
+ text
+(1 row)
+{% endhighlight %}
+
+*We cannot chain `->>` since it returns a TEXT.*
+
++ `#>` and `#>>` receives a path as an Array and returns a JSON and
+TEXT respectively. Both return `NULL` if the path cannot be found.
+
+{% highlight sql %}
+SELECT '{"state": {"name": "abc"} }'::json#>ARRAY['country', 'name'];
+
+?column?
+----------
+ "abc"
+(1 row)
+
+SELECT pg_typeof(
+  '{"state": {"name": "abc"} }'::json#>ARRAY['country','name']
+);
+
+ pg_typeof
+-----------
+ json
+(1 row)
+
+SELECT '{"state": {"name": "abc"} }'::json#>>ARRAY['country', 'name'];
+
+?column?
+----------
+ abc
+(1 row)
+
+SELECT pg_typeof(
+  '{"state": {"name": "abc"} }'::json#>ARRAY['country','name']
+);
+
+ pg_typeof
+-----------
+ text
+(1 row)
+{% endhighlight %}
+
+### JSONB
+
+`JSONB` adds new operators on top of the existing ones for the `JSON` datatype.
+
++ `@>` and `<@` operators performs like the ones we saw for the Array datatype.
+The first checks if the left JSONB contains the one in the right and the latter
```

Comment:
I will love if I had the same explanation (maybe partial or another) in the Array operator topic: who is compared with what.